engineering: Add GCP projects overview (#1413)
sourcegraph/about: handbook/engineering/environments.md

# Environment

This page describes our different cloud environments.

## Google Cloud

We utilize multiple Google Cloud projects and folders to organize our workloads and manage access control.

### Root Projects

- **Sourcegraph**: Services for sourcegraph.com.

> - **sourcegraph-code-intel**: Services for Code Intel code execution.

Reviewer: I think this should document why these exist. For example, we created this for isolation purposes.

Author: I can update the main description to explain that as its a general attribute more than a per-project one.

Author: I believe this should be covered by "We utilize multiple Google Cloud projects and folders to organize our workloads and manage access control." in the GCP section.

Author: @nicksnyder Updated, let me know your thoughts.

- **sourcegraph-managed-**: There are multiple `-managed` projects, one for each our managed deployments.
- **TelligentSourcegraph**: Data pipeline and storage for pings and Cloud event logging.
- **Sourcegraph Auxiliary**: Testing clusters and deployments.
- **Universities**: Sourcegraph instances for universities. (Deprecated)

> - **sourcegraph-interviews**: Shared project for interviews.

Reviewer: Fyi, I created this for the distribution team kubernetes coding exercise (so we can create a cluster for the candidate) but it hasn't really been used. lmk if you think we should keep this.

Author: I think we can keep it for now, Ill check what is running in it and the test to ensure its still useful, but in general I like providing a test environment for interviews.

> - **sourcegraph-calend**: Owned by @sqs.

Reviewer: What is it?

> - **sourcegraph-orgtool**: Owned by @sqs.

Reviewer: What is it?

### Folders

> Dynamic or temporary projects are created in the folders listed below.

Reviewer: ```suggestion

Dynamic or temporary projects are created in the folders listed below.
```

# Environment

This page describes our different cloud environments.

## Google Cloud

> We utilize multiple Google Cloud projects and folders to organize our workloads and manage access control.

Reviewer 2: Does someone need to request access or does everyone have access to these projects? 
If you need to request access, it might be helpful to have a note here on where to ask.

Author: Right now, it has to be requested, once we have it sorted out, I would like to for all engineers to have at least view access to most the projects.

### Root Projects

- **Sourcegraph**: Services for sourcegraph.com.
- **sourcegraph-code-intel**: Services for Code Intel code execution.
- **sourcegraph-managed-**: There are multiple `-managed` projects, one for each our managed deployments.
- **TelligentSourcegraph**: Data pipeline and storage for pings and Cloud event logging.
- **Sourcegraph Auxiliary**: Testing clusters and deployments.
- **Universities**: Sourcegraph instances for universities. (Deprecated)
- **sourcegraph-interviews**: Shared project for interviews

### Folders
Dynamic or temporary projects are created in the folders listed below.

> - **Engineers Projects**: Contains multiple engineers project. All projects must be prefixed with `$name-`.

Reviewer 3: Do we have this folder today? Because I can't see it... although I also can't see most of projects listed here 😂

Author: I created it yesterday, I still have to sort out permissions, most projects/folders and even the org are not seen by members right now.

Reviewer 4: If i'm understanding this right:

```suggestion
- **Engineers Projects**: Contains projects used by individual engineers. All projects must be prefixed with `$name-` (the name of the owner).
```

# Environment

This page describes our different cloud environments.

## Google Cloud

We utilize multiple Google Cloud projects and folders to organize our workloads and manage access control for our engineers, as well as limit the scope of roles and service accounts across projects.

### Root Projects

> These projects contain per-project permissions.

Reviewer: ```suggestion

These projects contain per-project permissions.
```

> These projects contain per-project permissions.

Reviewer: Can you document what "per-project" permissions means?

- **TelligentSourcegraph**: Data pipeline and storage for pings and Cloud event logging
- **Universities**: Sourcegraph instances for universities
- **sourcegraph-interviews**: Shared project for interviews

### Folders

> Dynamic or temporary projects are created in the folders listed below.

Reviewer: Can you document what it means for a project to be "dynamic" or "temporary"?

Author: Ill remove this, it does not make sense anymore

> ### Engineers Projects

Reviewer: ```suggestion

### Engineers Projects
```

Contains projects used by individual engineers. Engineers are expected to remove all their resources once they are done testing. All projects must be prefixed with `$name-` (the name of the owner).

> ### Sourcegraph Cloud

Reviewer: ```suggestion

### Sourcegraph Cloud
```

Sourcegraph Cloud projects.

- **Sourcegraph**: Services for sourcegraph.com
- **Sourcegraph Auxiliary**: Testing clusters, deployments and VMs

> - **sourcegraph-code-intel**: Services for Code Intel code execution

Reviewer: ```suggestion
- **sourcegraph-code-intel**: Services for Code Intel code execution that are separated from our production project for extra isolation.
```

@efritz Can you elaborate why we want this extra isolation?

Author: Roles in GCP grant permissions across the entire project, so if we create a role that has DB Admin permissions, it has admin to all databases in that project. GCP Recommends splitting into project, I believe this information might be a bit redundant here given the first section outlines:
> as well as limit the scope of roles and service accounts across projects.

Reviewer 5: Here or in the PR? I think @pecigonzalo, @slimsag, and @chayim could answer this question just as well. I think the main driver was wanting extra isolation to prevent root escalation opportunities from having access to the same GCP APIs in the project that runs our cloud cluster. We also thought it would be nice to separate for better cost accounting so we can determine how much this effort is actually affecting our compute budget.

Reviewer: I want to specifically document why code intel is a separate project here on this page since it is the only production service that is separate from the Sourcegraph project. Can either of you push a commit to document this?

Author: Will do

- **Sourcegraph CI**: Services for our CI cluster and temporary CI resources

> ### Other Projects

Reviewer: ```suggestion

### Other Projects
```
